pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+
+        pbft_msg msg;
+        if (bzn_envelope envelope{prepared_proof.pre_prepare()}; /*!envelope.ParseFromString(prepared_proof.pre_prepare()) ||*/ !this->is_peer(envelope.sender()) || !msg.ParseFromString(envelope.pbft()))
+        {
+            return false;
+        }
+        sequence_request_pairs.insert(std::make_pair(msg.sequence(), msg.request_hash()));
+    }
+    return true;
+}
+
+/**
+ * A valid newview message must contain a complete set of pre prepares, that is
+ * there must be one pre prepare for each sequence.
+ * TODO: this method needs a better name
+ * @param newview_msg
+ * @return true if there are no missing pre-prepares based on sequence numbers of each pre-prepare, false otherwise.
+ */
+bool
+pbft::pre_prepares_contiguous(uint64_t latest_sequence, const pbft_msg& newview_msg)
+{
+    if (0 == newview_msg.pre_prepare_messages_size())
+    {
+        return true;
+    }
+
+    auto extract_sequence = [](const bzn_envelope pre_prepare_envelope)->uint64_t
+            {
+                pbft_msg pre_prepare_message;
+                pre_prepare_message.ParseFromString(pre_prepare_envelope.pbft());
+                return pre_prepare_message.sequence();
+            };
+
+    // We are assuming that the pre prepare messages are stored sequentially by sequence number in newview_msg.pre_prepare_messages
+    uint64_t last_sequence{extract_sequence(newview_msg.pre_prepare_messages(0))};
+    for (int i{1}; i < newview_msg.pre_prepare_messages_size(); ++i)
+    {
+        uint64_t current_sequence{extract_sequence(newview_msg.pre_prepare_messages(i))};
+        if (last_sequence + 1 != current_sequence)
+        {
+            return false;
+        }
+        last_sequence = current_sequence;
+    }
+    return last_sequence==latest_sequence;
+}
+
+// prepared_proofs is not a member of a new_view message. You need to
+// iterate over all the prepared_proofs in all the viewchange messages
+// in the new_view.
+uint64_t
+pbft::last_sequence_in_newview_preprepare_messages(const pbft_msg &newview)
+{
+    uint64_t last_sequence{0};
+    for(int i{0}; i < newview.viewchange_messages_size(); ++i)
+    {
+        bzn_envelope viewchange_envelope = newview.viewchange_messages(i);
+        pbft_msg viewchange;
+        viewchange.ParseFromString(viewchange_envelope.pbft());
+
+        for(int j{0}; j < viewchange.prepared_proofs_size(); ++j)
+        {
+            prepared_proof proof = viewchange.prepared_proofs(j);
+            for(int k{0}; k < proof.prepare_size(); ++k)
+            {
+                bzn_envelope prepare = proof.prepare(k);
+                pbft_msg idfk;
+                idfk.ParseFromString(prepare.pbft());
+                last_sequence = std::max(last_sequence, idfk.sequence());
+            }
+        }
+    }
+    return last_sequence;
+}
+
+bool
+pbft::is_valid_newview_message(const pbft_msg& thiers, const bzn_envelope& original_thiers) const
```

Comment:
Hate to be so pedantic, but "theirs" not "thiers"

Reply:
CRAP! I hate it when I do that! Fixed both arguments.